haystack_experimental/components/tools/tool_invoker.py

```diff
@@ -0,0 +1,204 @@
+        return f"The weather in {city} is 20 degrees."
+
+    parameters = {"type": "object",
+                "properties": {"city": {"type": "string"}},
+                "required": ["city"]}
+
+    tool = Tool(name="weather_tool",
+                description="A tool to get the weather",
+                function=dummy_weather_function,
+                parameters=parameters)
+
+    # Usually, the ChatMessage with tool_calls is generated by a Language Model
+    # Here, we create it manually for demonstration purposes
+    tool_call = ToolCall(
+        tool_name="weather_tool",
+        arguments={"city": "Berlin"}
+    )
+    message = ChatMessage.from_assistant(tool_calls=[tool_call])
+
+    # ToolInvoker initialization and run
+    invoker = ToolInvoker(tools=[tool])
+    result = invoker.run(message=message)
+
+    print(result)
+    ```
+
+    ```
+    >>  {
+    >>      'tool_messages': [
+    >>          ChatMessage(
+    >>              _role=<ChatRole.TOOL: 'tool'>,
+    >>              _content=[
+    >>                  ToolCallResult(
+    >>                      result='"The weather in Berlin is 20 degrees."',
+    >>                      origin=ToolCall(
+    >>                          tool_name='weather_tool',
+    >>                          arguments={'city': 'Berlin'},
+    >>                          id=None
+    >>                      )
+    >>                  )
+    >>              ],
+    >>              _meta={}
+    >>          )
+    >>      ]
+    >>  }
+    ```
+    """
+
+    def __init__(self, tools: List[Tool], raise_on_failure: bool = True):
+        """
+        Initialize the ToolInvoker component.
+
+        :param tools:
+            A list of tools that can be invoked.
+        :param raise_on_failure:
+            If True, the component will raise exceptions in case the tool is not found or invocation fails.
+            If False, the component will return a ChatMessage from the tool, wrapping the error message.
+        """
+        if not tools:
+            raise ValueError("ToolInvoker requires at least one tool to be provided.")
+
+        self.tools = tools
+        self._tools_with_names = {tool.name: tool for tool in tools}
+        self.raise_on_failure = raise_on_failure
+
+    @staticmethod
+    def _convert_tool_result_to_string(result: Any) -> str:
+        """
+        Converts the tool invocation result to a string.
+
+        :param result:
+            The tool result.
+        :returns:
+            The string representation of the tool result.
+        """
+        try:
+            return json.dumps(result)
+        except TypeError as e:
+            logger.warning(
+                "Failed to convert tool result to string using `json.dumps`. Error: {error}. Falling back to `str`.",
+                error=e,
+            )
+            return str(result)
+
+    @component.output_types(tool_messages=List[ChatMessage])
+    def run(self, message: ChatMessage):
```

Comment:
Why just a single chat message? It'd be more useful to use the same modality of the generator's outputs, i.e., a list of chat messages are accepted and each tool call message is replaced with the tool call result(s).

That way, you don't need another component in-between the generator and the invoker to preprocess the former's outputs.

Reply:
In the most common scenarios, you have a _single_ message from assistant with one or more tool calls and you then execute these invocations, returning a list of (one or more) messages.
I'm quite confident that this design can fit into scenarios like the one outlined by [my old experiments] (https://colab.research.google.com/drive/1ZxUiTsYDBckktywK65wYazEZCqVVyHlA?usp=sharing) with some small changes.

Our current [`OpenAIFunctionCaller`](https://github.com/deepset-ai/haystack-experimental/blob/main/haystack_experimental/components/tools/openai/function_caller.py) takes `List[ChatMessage] ` as input but only processes the first one.

So if we take a list of messages with tool calls as input we can alternatively choose to:
- process the first or the last one
- return a list of list of messages
- (flat the previous list of lists)

Happy to discuss...

Reply:
We discussed this offline and there could be several different sensible ways to structure inputs and outputs for this component.

During the discussion, I liked the idea of accepting a list of messages, processing only those with tool calls, and returning a list of (tool) messages. So I modified the implementation in this direction.

However, after merging this PR I will play with the current implementation and see if it is ergonomic and fits well with common use cases. We can then modify it later.